design-proposals/backend-storage-clone-on-migration/backend-storage-clone-on-migration.md

```diff
@@ -0,0 +1,106 @@
+# Overview
+This proposal is about supporting the migration of VMs that use persistent features, such as persistent vTPM, with any [storage class](https://kubernetes.io/docs/concepts/storage/storage-classes/).
+
+To make this document as simple as possible, let's consider that storage in Kubernetes can either be filesystem (FS) or
+block mode, and either ReadWriteOnce (RWO) or ReadWriteMany (RWX) access mode, with storage classes supporting any
+number of combinations of both.
+
+As of KubeVirt 1.3, Here is how each of the 4 combinations is supported by the backend-storage for persistent VM state:
+- RWO FS: supported by backend-storage but makes VMs non-migratable. **This design proposal is about removing that limitation.**
+- RWX FS: preferred combination, supported by backend-storage and compatible with VM migrations
+- RWO block: not supported
+- RWX block: not supported
+
+This proposal is about switching to a different model which would instead create a new backend-storage PVC on migration, enabling RWO FS.  
+It is important to note that RWO FS is supported by most, if not all, Kubernetes storage classes out there, as it is needed for basic pod functionalities.  
+Implementing this proposal will therefore enable the of persistent features on any clusters, no matter which storage classes are present.
+
+## Alternative approaches
+
+### Direct block support in upstream projects
+
+An alternative to this design would be for upstream components (libvirt - qemu - swtpm) to add block storage support for TPM/EFI.  
+KubeVirt could then create one backend-storage PVC per feature (EFI/TPM) per VM and pass them directly to libvirt.  
+That alternative approach would enable RWX block backend-storage (as opposed to RWO FS in this proposal).  
+
+Pros:
+- Less code to write and maintain in KubeVirt
+- No need to duplicate backend-storage PVCs on migrations and delete the correct one after migration success/failure
+- Libvirt gains a feature
+
+Cons:
+- Relies on third party projects adding and maintaining options to store config files into a block devices without filesystems
+- Requires multiple PVCs per VM (one per persistence-enabled feature)
+- We would have to either
+  - Maintain 2 different backend-storage implementations (the existing RWX FS one and this RWX block one)
+  - Deprecate RWX FS and require RWX block to enable backend-storage
```

Comment:
With the primary design, we keep support for RWX FS and extend it to support RWO FS. Since both use virtually the same codepath, it's just one thing to maintain.
However, RWX block would need a completely different approach, so if we want to avoid maintaining 2 different ways to do the same thing we'd have to deprecate RWX FS.

Reply:
I don't know the code, but ideally, by delegating the trouble to libvirt, the only difference between RWX FS and RWX Block would be passing a block device to libvirt instead of a filesystem path.

In the primary implementation, we need to add code to synchronize between source and destination during success and various modes of failures.

Reply:
> I don't know the code, but ideally, by delegating the trouble to libvirt, the only difference between RWX FS and RWX Block would be passing a block device to libvirt instead of a filesystem path.

Yes, well multiple block devices. That means changing every function that touches backend-storage to handle multiple PVCs.
It would also require changing the volumes and mounts we pass to virt-launcher. Finally the libvirt options would indeed look different.
That is a non-trivial amount of differences.

> In the primary implementation, we need to add code to synchronize between source and destination during success and various modes of failures.

Yes, before we can mark a (RWO) migration as completed we have to add a label to a PVC and delete another one. It is an extra step, and we have to make sure it's done right, but it's not much code at all.